SignalMessaging/utils/ThreadUtil.m

```diff
@@ -527,137 +527,94 @@ + (void)ensureUnreadIndicator:(ThreadDynamicInteractions *)dynamicInteractions
-            dynamicInteractions.firstUnseenInteractionTimestamp = nil;
-        } else if (dynamicInteractions.hasMoreUnseenMessages) {
-            NSMutableSet<NSData *> *missingUnseenSafetyNumberChanges = [NSMutableSet set];
-            for (TSInvalidIdentityKeyErrorMessage *safetyNumberChange in blockingSafetyNumberChanges) {
-                BOOL isUnseen = safetyNumberChange.timestampForSorting
-                    >= dynamicInteractions.firstUnseenInteractionTimestamp.unsignedLongLongValue;
-                if (!isUnseen) {
-                    continue;
-                }
-                BOOL isMissing
-                    = safetyNumberChange.timestampForSorting < interactionAfterUnreadIndicator.timestampForSorting;
-                if (!isMissing) {
-                    continue;
-                }
-
-                NSData *_Nullable newIdentityKey = safetyNumberChange.newIdentityKey;
-                if (newIdentityKey == nil) {
-                    OWSFail(@"Safety number change was missing it's new identity key.");
-                    continue;
-                }
+    __block BOOL hasMoreUnseenMessages = NO;
+    [threadMessagesTransaction
+        enumerateRowsInGroup:thread.uniqueId
+                 withOptions:NSEnumerationReverse
+                  usingBlock:^(
+                      NSString *collection, NSString *key, id object, id metadata, NSUInteger index, BOOL *stop) {
+                      if (![object isKindOfClass:[TSInteraction class]]) {
+                          OWSFail(@"Expected a TSInteraction: %@", [object class]);
+                          return;
+                      }
+
+                      TSInteraction *interaction = (TSInteraction *)object;
+
+                      if (interaction.isDynamicInteraction) {
+                          // Ignore dynamic interactions, if any.
+                          return;
+                      }
+
+                      if (interaction.timestampForSorting < firstUnseenInteractionTimestamp.unsignedLongLongValue) {
+                          // By default we want the unread indicator to appear just before
+                          // the first unread message.
+                          *stop = YES;
+                          return;
+                      }
+
+                      visibleUnseenMessageCount++;
+
+                      interactionAfterUnreadIndicator = interaction;
+
+                      if (visibleUnseenMessageCount + 1 >= maxRangeSize) {
+                          // If there are more unseen messages than can be displayed in the
+                          // messages view, show the unread indicator at the top of the
+                          // displayed messages.
+                          *stop = YES;
+                          hasMoreUnseenMessages = YES;
+                      }
+                  }];
+
+    if (!interactionAfterUnreadIndicator) {
+        // If we can't find an interaction after the unread indicator,
```

Comment:
This comment seems out of date.

Reply:
It mostly still applies, but I've reworded it a little bit for clarity.

* We want to have "unread indicator continuity", e.g. we want to leave the unread indicator "in place" even after we've marked messages after it as read.  We persist `firstUnseenInteractionTimestamp` to accomplish this.
* `firstUnseenInteractionTimestamp` can get stale - all messages after the unread indicator may have disappeared or been manually deleted, etc.  In that case, we don't want to show the unread indicator.